Fix graph labels date

AlbionaHoti opened the pull request:
I found the ```GraphsContainer.jsx``` file where the data was showing up, my solution was to update the date when there was no data, the moment library showed the default date which is 01 Jan 1970 - So I checked if the date is earlier than the 1975 and I added the current date to the view.

@sarah-clements I updated the branch and the name.

sarah-clements commented:
>I found the GraphsContainer.jsx file where the data was showing up, my solution was to update the date when there was no data, the moment library showed the default date which is 01 Jan 1970 - So I checked if the date is earlier than the 1975 and I added the current date to the view.

I think this is a good first attempt but when I look at this link it is only updating the date on the top graph: http://localhost:5000/perf.html#/graphs?highlightAlerts=1&series=mozilla-central,1834678,1,10&series=mozilla-central,1955233,1,10&series=mozilla-central,2110541,1,10&series=mozilla-central,2117747,1,10&series=mozilla-central,1834673,1,10&series=mozilla-central,1955223,1,10&series=mozilla-central,2110530,1,10&series=mozilla-central,2117883,1,10&series=mozilla-central,1834668,1,10&series=mozilla-central,2008512,1,10&series=mozilla-central,2110577,1,10&series=mozilla-central,2117729,1,10&timerange=86400

Rather than hard-coding a specific date format ('1975-01-01'), I suggest a solution that would where you check if data is available. 

>@sarah-clements I updated the branch and the name.
Please also change your first commit so it includes the bug number: `Bug 1584135 - Fix graph labels data`. We have a bot that automatically updates the bugzilla bug when new prs are opened that reference the bug number.

AlbionaHoti commented:
> 
> 
> > I found the GraphsContainer.jsx file where the data was showing up, my solution was to update the date when there was no data, the moment library showed the default date which is 01 Jan 1970 - So I checked if the date is earlier than the 1975 and I added the current date to the view.
> 
> I think this is a good first attempt but when I look at this link it is only updating the date on the top graph: http://localhost:5000/perf.html#/graphs?highlightAlerts=1&series=mozilla-central,1834678,1,10&series=mozilla-central,1955233,1,10&series=mozilla-central,2110541,1,10&series=mozilla-central,2117747,1,10&series=mozilla-central,1834673,1,10&series=mozilla-central,1955223,1,10&series=mozilla-central,2110530,1,10&series=mozilla-central,2117883,1,10&series=mozilla-central,1834668,1,10&series=mozilla-central,2008512,1,10&series=mozilla-central,2110577,1,10&series=mozilla-central,2117729,1,10&timerange=86400
> 
> Rather than hard-coding a specific date format ('1975-01-01'), another solution is to check if data is available.

To check if data is available when there is no available date, moment library gives the default date which is 1970, 1st January.

This seems to be available but the req was to set up the current date, not the 1st of Jan. I will check if I can make another solution to remove the hard-coding date.
> 
> > @sarah-clements I updated the branch and the name.
> 
> Please also change your first commit so it includes the bug number: `Bug 1584135 - Fix graph labels data`. We have a bot that automatically updates the bugzilla bug when new prs are opened that reference the bug number.

Ok, perfect.

AlbionaHoti commented:
@sarah-clements I did a change on the code, ```moment(x).diff(moment(['1975-01-01']), 'years') < 10``` I had to use this date so I could do a comparison, the date that we get when there is no data is ```1970-01-01``` and is a valid date. 

What do you think? Is it okay to let the current date, when there is no new data?

sarah-clements commented on ui/perfherder/graphs/GraphsContainer.jsx:
```diff
@@ -430,6 +434,11 @@ class GraphsContainer extends React.Component {
               />
               <VictoryAxis
                 tickCount={6}
+                tickFormat={x =>
```
tickFormat is now listed twice due to changes someone else made, so your function is not being applied (the set tickFormat is overriding it).

sarah-clements commented on ui/perfherder/graphs/GraphsContainer.jsx:
```diff
@@ -321,7 +321,11 @@ class GraphsContainer extends React.Component {
               <VictoryAxis
                 tickCount={10}
-                tickFormat={x => moment.utc(x).format('MMM DD')}
+                tickFormat={x =>
```
The problem I see with this approach is that we're relying on a default date that the graphing library is providing when there isn't data. This default value could change at any time and we may not know about it. It's not actually a default date provided by moment - x is a Date object that is being formatted to a localized time by moment. I think a more stable approach is to determine where the source of data is coming from for each graph and to test for it - if it's empty, show the current day's date and if it isn't, show the date provided by x. 

Also, the same code is being written twice so this is something that you'd want to write once and store in a variable or function.

AlbionaHoti replied:
I will put the code into one func, or var and use it in both places. Thank you for the feedback :)

sarah-clements commented:
> @sarah-clements I did a change on the code, `moment(x).diff(moment(['1975-01-01']), 'years') < 10` I had to use this date so I could do a comparison, the date that we get when there is no data is `1970-01-01` and is a valid date.
> 
> What do you think? Is it okay to let the current date, when there is no new data?

The problem I see with this approach is that we're relying on a default date that the graphing library is providing when there isn't data. This default value could change at any time and we may not know about it. It's not actually a default date provided by moment - x is a Date object that is being formatted to a localized time by moment. I think a more stable approach is to determine where the source of data is coming from for each graph and to test for it - if it's empty, show the current day's date and if it isn't, show the date provided by x.  Another solution is to not show the graphs if there isn't data, but to show a message instead that explains to the user that there isn't data for that selected date range.

AlbionaHoti commented:
> 
> 
> > @sarah-clements I did a change on the code, `moment(x).diff(moment(['1975-01-01']), 'years') < 10` I had to use this date so I could do a comparison, the date that we get when there is no data is `1970-01-01` and is a valid date.
> > What do you think? Is it okay to let the current date, when there is no new data?
> 
> The problem I see with this approach is that we're relying on a default date that the graphing library is providing when there isn't data. This default value could change at any time and we may not know about it. It's not actually a default date provided by moment - x is a Date object that is being formatted to a localized time by moment (with your code). I think a more stable approach is to determine where the source of data is coming from for each graph and to test for it - if it's empty, show the current day's date and if it isn't, show the date provided by x. Another solution is to not show the graphs if there isn't data but to show a message instead that explains to the user that there isn't data for that selected date range.
> 
> If you have any questions you can also ask in the IRC channel :)

> My solution based on feedback: 

- I created the function ```checkDate(x)``` -> In here I check the prop testData - if the visible property is false or true, based on this data we know that we can display the data in the graph, so I check if the ```visible``` prop is true, then continue to add the data of the ```testData```, if not, use the current date. 
- When there is the ```testData``` but the 'data' property array is empty, it shows the base date of Jan 01.

- Let me know your thoughts. @sarah-clements

sarah-clements commented on ui/perfherder/graphs/GraphsContainer.jsx:
```diff
@@ -213,6 +213,17 @@ class GraphsContainer extends React.Component {
+    );
+    return isVisible
+      ? moment.utc(x).format('MMM DD')
+      : moment.utc(new Date()).format('MMM DD');
```
This was a good attempt :)

We only disable a few select eslint rules but usually for specific reasons (like a poorly written rule). In this case, it's a hint that you are doing something that isn't best practice (`isVisible = true`). I suggest reading the rule for future reference :)

So this approach doesn't quite make sense - we don't want to return the date based on the last value `isVisible` is set to - we want to make a quick determination about whether not we can consider `testData` to be empty or to not contain items that are visible. `filter` is a better method than `map` to use here. Sometimes it helps to write a function in pseudo-code to make sure the logic is correct.

Also I think if you do `moment().utc().format('MMM DD')` that achieves the same thing as `moment.utc(new Date()).format('MMM DD')`.

sarah-clements replied:
There are merge conflicts so you will need to rebase onto the latest master (don't do pull and merge).

AlbionaHoti replied:
@sarah-clements Thank you for your feedback, I see my mistakes now and tried to go through the prob in a better way, I think.

What I did is:
```
 checkDate = x => {
    let newArr = [];
    newArr = this.props.testData.filter(
      item => item.visible === true && item.data.length > 0,
    );

    return newArr.length > 0
      ? moment.utc(x).format('MMM DD')
      : moment.utc().format('MMM DD');
  };
```

- When we check cards, and visible prop is true, but the card data is empty, we show the current date. 
`Question: Do you want to show the current date in selected cards, with no data inside it? If we don't want to show the current date, but check only the visible prop is true, then the 01 Jan will be shown.`
- Do we have to check for the visible and data if they are true? 
- The solution I provide checks the visible prop and the data length for cards that are selected -> save items that have these two things, if we have data in the new array then show the date from the data, if not, show the current date.
- Please let me know if you have a better approach, thank you.
- I will not push the code yet.

sarah-clements replied:
This looks better. To answer your question, we definitely don't want to show the 01 Jan date at all. I'm assuming you tested this approach using the link in this [comment](https://github.com/mozilla/treeherder/pull/5440#issuecomment-538133224)?

A few nits:
- You should declare `newArr` once (instead of declaring it with an empty array and reassigning it) because `filter` will return an empty array if there aren't any items that meet the condition: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter
- Rename `newArr` to something meaningful like `graphData` or `data`
- I think you need to call `moment().utc().format('MMM DD')` - to get today's date. You can check the momentjs docs to be certain

AlbionaHoti replied:
- I tested this approach with the link in that comment.
- We don't show the 01 Jan then, the current date will be shown when visible is true but there is no data.
- Thank you for your feedback on not declaring the `newArr`
I created a const and attach to it the calculation.
```
  checkDate = x => {
    const graphData = this.props.testData.filter(
      item => item.visible === true && item.data.length > 0,
    );

    return graphData.length > 0
      ? moment.utc(x).format('MMM DD')
      : moment().format('MMM DD');
  };
```
- Related to the moment library:
	* If we want to show the users local time from moment library, then we can use the `moment().format('MMM DD')` - While for the UTC - the docs says:  all display methods will display in UTC time instead of local time.
	* The difference:
		``` 
		moment().format();     // 2013-02-04T10:35:24-08:00
		moment.utc().format(); // 2013-02-04T18:35:24+00:00 
		```
	* Which one to use? I think using either option would work because we are not showing the hour, what do you think?

sarah-clements replied:
ah ok, we should only being showing utc date - not localized - so let's go with that format (it does matter because at some point local date for PST (for example) could be 2013-02-04 but be 2013-02-05 UTC).

AlbionaHoti replied:
@sarah-clements Hi, can you give me a hint related to the failed `continuous-integration/travis-ci/`

sarah-clements replied:
These errors are unrelated to what you've done. We need to revert someone else's commit that caused issues. I'll let you know when it's fixed.

sarah-clements replied:
This has been fixed now.

sarah-clements commented:
Can you remove the "Merge branch 'master' into change-graph-labels" commit ad27ad0? You should only use rebase not pull/merge. Here's a conversation a few days ago between kwierso and mellinay that might explain it a bit clearer: https://mozilla.logbot.info/treeherder/20191007#c16674431
(I've highlighted where to start reading). Let me know if that helps :)

AlbionaHoti commented:
> 
> 
> Can you remove the "Merge branch 'master' into change-graph-labels" commit [ad27ad0](https://github.com/mozilla/treeherder/commit/ad27ad0a0afb2738516cf8243ac6f71571709224)? You should only use rebase not pull/merge. Here's a conversation a few days ago between kwierso and mellinay that might explain it a bit clearer: [mozilla.logbot.info/treeherder/20191007#c16674431](https://mozilla.logbot.info/treeherder/20191007#c16674431)
> (I've highlighted where to start reading). Let me know if that helps :)

I will create a new pr with a new branch with our last changes if that's okay? @sarah-clements

sarah-clements commented:
There's no need to open another branch and it's preferable to keep this one open because the entire history of the review is in this pull request. All you need to do is look up how to remove the last commit and then use rebase to get your branch up to date. If you get stuck, you can ask for questions on the channel. :)

AlbionaHoti commented:
Hi @sarah-clements I am trying to fix this, it says 'the repo that submitted this pr has been deleted'. I deleted the repo I forked and where I added the changes for this pr. Is there another option on how to move forward with this?
